### DISCUSSION TITLE
Get prices from graphQL is null

### DISCUSSION THREAD

**Discussion by tarikrital** [swellstores/community#527] [Help] [answered]
Welcome,

Today while trying to access the prices using a customer with Wholesale group prices is null? (I'm using horizon theme)

![image](https://github.com/swellstores/community/assets/52907282/4c00d017-d2ea-4c3b-8e6f-86d106bfc7f3)

The problem with products not Customer, I was able to get the Customer using swell-node

![image](https://github.com/swellstores/community/assets/52907282/2eaeb15e-4e1a-48c5-9ba5-e808502a819c)

**Comment by tarikrital** [comment]
ℹ️  I was able to fix the issue using swell-node but the question here is for GQL.

**Reply by logeshswell** [reply]
Can you share the graphQL query along with the sample product and customer Ids for reference?

**Reply by logeshswell** [reply]
`prices` value will be set only when you add price rules for the given variant.

**Reply by logeshswell** [reply]
<img width="910" alt="Screenshot 2024-05-02 at 5 26 11 PM" src="https://github.com/swellstores/community/assets/100766131/29f3e3d8-3287-438e-86de-7c9bd5fd38f9">

I don't see any price rules set for the given variant. Hence, you see `null` value being returned in graphQL.

**Reply by tarikrital** [reply]
The Product id is : `65c13f1c1c0dbd001200c9c4`

I added the rule to variant too but this time it shows 0 in the price.
![image](https://github.com/swellstores/community/assets/52907282/919be9d9-0d02-4b50-819a-815ff413463b)

In contrast on products list its shows the normal price even the user connected with group targeted by the price rule.

![image](https://github.com/swellstores/community/assets/52907282/1af1dcd0-bd29-4db6-bbb8-e406fe975306)

**Comment by tarikrital** [comment]
I found [this](https://github.com/orgs/swellstores/discussions/275#discussioncomment-6232569) from @ericingram.

**Comment by logeshswell** [comment, marked as answer]
I confirm that it's not a bug in GraphQL API.

GraphQL API is working fine as follows:
<img width="1478" alt="Screenshot 2024-05-02 at 6 54 10 PM" src="https://github.com/swellstores/community/assets/100766131/04f4ac5e-d1b1-4a5d-ab34-3254040ec354">

As you can see, it returns `variants.results.prices` value correctly when the wholesale user is logged into the current session.

Secondly, in the horizon theme, the price is shown as $19.99 for the product.

When I added the product with quantity = 1, the cart contained the same product price.
<img width="1511" alt="Screenshot 2024-05-02 at 6 51 12 PM" src="https://github.com/swellstores/community/assets/100766131/e1da3780-d3c6-4580-b026-f86a5f3acdf5">

When I increased the quantity to 2, the cart contained $20 (= 2 * $10).
<img width="1512" alt="Screenshot 2024-05-02 at 6 51 22 PM" src="https://github.com/swellstores/community/assets/100766131/fcfa590b-d239-4ecc-9a93-bc7e92875de1">

**Comment by tarikrital** [comment]
Thank you, @logeshswell.

Excellent confirmation, it seems a problem in Horizon theme which is abandoned from Swell because it shows on wrong price I think also on the images you attached and products list too.

**Reply by logeshswell** [reply]
No, it's not a bug in horizon theme as well. 

`variants.results.price` value is displayed in the product details page because the product (or variant) can have multiple pricing rules based on a customer group and min/max quantity. We can't show the price value without the input of required quantity value.

But, you can extend it as per your usecase.

**Reply by tarikrital** [reply]
> No, it's not a bug in horizon theme as well.

In the product details page also shows null.

**Comment by tarikrital** [comment]
What was your headers keys? here shows null!

![image](https://github.com/swellstores/community/assets/52907282/3b036b16-0627-4920-80a0-765275465bc6)

**Reply by logeshswell** [reply]
You should pass "x-session" header, which determines the customer account using which the products are accessed.

**Comment by tarikrital** [comment]
This the query in our project:
```

export const GetAllProductsDocument = `
    query getAllProducts($currency: String) {
  products(_currency: $currency) {
    count
    results {
      id
      name
      description
      slug
      price
      sale
      salePrice
      origPrice
      currency
      attributes
      categories {
        name
        slug
      }
      images {
        caption
        file {
          width
          height
          url
        }
      }
      options {
        id
        attributeId
        name
        inputType
        active
        required
        variant
        values {
          id
          name
          price
          description
        }
      }
      purchaseOptions {
        standard {
          price
          sale
          salePrice
          origPrice
        }
        subscription {
          active
        }
      }
      variants {
        results {
          name
          price
          origPrice
          prices {
            price
            discountPercent
            quantityMin
            quantityMax
            accountGroup
          }
          images {
            id
            caption
            file {
              url
              width
              height
            }
          }
          optionValueIds
          purchaseOptions {
            standard {
              price
              sale
              salePrice
              origPrice
              prices {
                price
                discountPercent
                quantityMin
                quantityMax
                accountGroup
              }
            }
          }
          stockLevel
          currency
        }
      }
    }
  }
}
    `;
```

**Comment by tarikrital** [comment]
Welcome @logeshswell,

What is the appropriate way to use user session in the products slugs static page [`[slug].tsx`](https://github.com/swellstores/horizon/blob/main/pages/products/%5Bslug%5D.tsx) with GQL?

Currently, I call a local API to get the session and it works in [`index.tsx`](https://github.com/swellstores/horizon/blob/main/pages/products/index.tsx) and returns the Price rule in the GQL result but not on the `[slug].tsx`!

**Reply by logeshswell** [reply]
You just need to pass `x-session` header in the GraphQL request.
Can you check the implementation [here](https://github.com/swellstores/horizon/blob/main/lib/shop/fetchingFunctions.ts) and try it out for slug.tsx?

**Reply by tarikrital** [reply]
I worked with that file, did many attempts and I modified it.

In step 1 on products page receives the session/cookie and returns **Price rule** for the user **group** but step 2 not:

Step 1:

```
export const getProductsList = async (
  categorySlug?: string,
  currency?: string,
) => {
   const response = await fetch(
      `${API_BASE_URL}/api/swell/client`,
    );
    const data = await response.json();
  const authClient = getClientWithSessionToken(data);

  const { products } = await authClient
    .getAllProducts({ currency })
    .then((response) => response.data);
```
Step 2:

```
export async function getProductBySlug(
  slug: string,
  options?: { currency?: string; locale?: string },
): Promise<ProductsPageProps> {
  const { currency, locale } = options ?? {};
     const authResponse = await fetch(
      `${API_BASE_URL}/api/swell/client`,
    );
    const data = await authResponse.json();
     console.log('getProductBySlug data client:', data);

  const authClient = getClientWithSessionToken(data);
const response = await authClient.getProduct({
    slug,
    currency,
    locale,
  });
  const { productBySlug: product } = response.data;
```

**Reply by logeshswell** [reply]
Got it, can you check the implementation of `authClient.getProduct`?

I can confirm that `getProductBySlug` query can also return the price rule for the user group as returned for Step-1.

**Reply by tarikrital** [reply]
It returns using Playground as we mentioned before but locally using Horizon theme design static page not!

![image](https://github.com/swellstores/community/assets/52907282/6ad765d1-7456-4ec4-beab-b6ea7d355594)

I modified the query locally and the generated file works on the Playground but the session is {} locally!

**Reply by logeshswell** [reply]
Please try to understand the internals of the Horizon theme and the static build and then figure out how to set the session correctly.
